File: frontend/src/types/report.ts
export type AnalysisStatus =
  | "pending"
  | "running"
  | "completed"
  | "failed";

export type PriorityRating =
  | "high"
  | "interesting_with_doubts"
  | "low"
  | "insufficient_data";

export type ConfidenceLevel = "high" | "medium" | "low";

export type EvidenceType =
  | "verified_fact"
  | "reasonable_inference"
  | "unverifiable";

export interface RiskItem {
  name: string;
  severity: "high" | "medium" | "low";
  evidence: string;
  explanation: string;
  what_to_validate: string;
  evidence_type: EvidenceType;
}

export interface AnalysisStatusResponse {
  analysis_id: string;
  status: AnalysisStatus;
  pipeline_step: string | null;
  company_name: string;
  website_url: string;
  started_at: string | null;
  completed_at: string | null;
  error_message: string | null;
}

export interface ReportResponse {
  report_id: string;
  analysis_id: string;
  company_name: string;
  website_url: string;
  factual_profile: Record<string, unknown>;
  business_model: Record<string, unknown>;
  traction_signals: Record<string, unknown>;
  risks: RiskItem[];
  references: Record<string, unknown>;
  conclusion: {
    priority_rating: PriorityRating;
    summary: string;
    key_questions: string[];
    confidence: ConfidenceLevel;
  };
  management: Record<string, unknown> | null;
  competitive_position: Record<string, unknown> | null;
  operational_signals: Record<string, unknown> | null;
  corporate_structure: Record<string, unknown> | null;
  executive_summary: {
    what_it_does: string | null;
    investment_snapshot: string[];
    key_risks: string[];
    final_verdict: string | null;
    final_score: number;
    score_label: string;
    priority_rating: PriorityRating;
    computed_data_confidence: "high" | "medium" | "low";
  } | null;
  priority_rating: PriorityRating;
  generated_at: string;
}

File: backend/app/schemas/report.py
from pydantic import BaseModel, field_validator
from uuid import UUID
from datetime import datetime
from typing import Optional, Any, List, Dict
from app.models.report import PriorityRating


class RiskItem(BaseModel):
    name: str
    severity: str  # high | medium | low
    evidence: str
    explanation: str
    what_to_validate: str
    evidence_type: str  # verified_fact | reasonable_inference | unverifiable

    @field_validator("what_to_validate", mode="before")
    @classmethod
    def normalize_what_to_validate(cls, v: Any) -> str:
        """Old Gemini runs (pre-JSON-mode) stored what_to_validate as a list[str].
        New runs with response_mime_type='application/json' correctly return a str.
        Normalize list → str so both formats deserialize without ValidationError.
        """
        if isinstance(v, list):
            return " ".join(str(item).strip() for item in v if item)
        return v


class ReportResponse(BaseModel):
    report_id: UUID
    analysis_id: UUID
    company_name: str
    website_url: str
    factual_profile: Dict[str, Any]
    business_model: Dict[str, Any]
    traction_signals: Dict[str, Any]
    risks: List[RiskItem]
    references: Dict[str, Any]
    conclusion: Dict[str, Any]
    management: Optional[Dict[str, Any]] = None
    competitive_position: Optional[Dict[str, Any]] = None
    operational_signals: Optional[Dict[str, Any]] = None
    corporate_structure: Optional[Dict[str, Any]] = None
    executive_summary: Optional[Dict[str, Any]] = None
    priority_rating: PriorityRating
    generated_at: datetime

    model_config = {"from_attributes": True}
